test_with_globbing_patterns_with_single_image sometimes fails on Arm

`test_with_globbing_patterns_with_single_image` **sometimes** fails on Arm (aarch64, armv7 and armv6). It looks like a race condition because the `nii` file is not available.

Nilearn version: 0.7.1

### Expected behavior
The test should be reliable.

### Actual behavior
When the test fails, the log is:
```
[  304s] =================================== FAILURES ===================================
[  304s] ________________ test_with_globbing_patterns_with_single_image _________________
[  304s] [gw10] linux -- Python 3.9.4 /usr/bin/python3.9
[  304s] 
[  304s]     def test_with_globbing_patterns_with_single_image():
[  304s]         # With single image
[  304s]         data_4d = np.zeros((40, 40, 40, 3))
[  304s]         data_4d[20, 20, 20] = 1
[  304s]         img_4d = nibabel.Nifti1Image(data_4d, affine=np.eye(4))
[  304s]         multi_pca = MultiPCA(n_components=3)
[  304s]     
[  304s]         with write_tmp_imgs(img_4d, create_files=True, use_wildcards=True) as img:
[  304s]             input_image = _tmp_dir() + img
[  304s] >           multi_pca.fit(input_image)
[  304s] 
[  304s] nilearn/decomposition/tests/test_multi_pca.py:184: 
[  304s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  304s] nilearn/decomposition/base.py:418: in fit
[  304s]     data = mask_and_reduce(
[  304s] nilearn/decomposition/base.py:172: in mask_and_reduce
[  304s]     data_list = Parallel(n_jobs=n_jobs)(
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:1044: in __call__
[  304s]     while self.dispatch_one_batch(iterator):
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:859: in dispatch_one_batch
[  304s]     self._dispatch(tasks)
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:777: in _dispatch
[  304s]     job = self._backend.apply_async(batch, callback=cb)
[  304s] /usr/lib/python3.9/site-packages/joblib/_parallel_backends.py:208: in apply_async
[  304s]     result = ImmediateResult(func)
[  304s] /usr/lib/python3.9/site-packages/joblib/_parallel_backends.py:572: in __init__
[  304s]     self.results = batch()
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:262: in __call__
[  304s]     return [func(*args, **kwargs)
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:262: in <listcomp>
[  304s]     return [func(*args, **kwargs)
[  304s] nilearn/decomposition/base.py:211: in _mask_and_reduce_single
[  304s]     this_data = masker.transform(img, confound)
[  304s] nilearn/input_data/multi_nifti_masker.py:344: in transform
[  304s]     return self.transform_single_imgs(imgs)
[  304s] nilearn/input_data/nifti_masker.py:443: in transform_single_imgs
[  304s]     data = self._cache(filter_and_mask,
[  304s] /usr/lib/python3.9/site-packages/joblib/memory.py:352: in __call__
[  304s]     return self.func(*args, **kwargs)
[  304s] nilearn/input_data/nifti_masker.py:56: in filter_and_mask
[  304s]     imgs = _utils.check_niimg(imgs, atleast_4d=True, ensure_ndim=4)
[  304s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  304s] 
[  304s] niimg = '/tmp/nilearn_h4rbgbk1.nii', ensure_ndim = 4, atleast_4d = True
[  304s] dtype = None, return_iterator = False, wildcards = True
[  304s] 
[  304s]     def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
[  304s]                     return_iterator=False,
[  304s]                     wildcards=True):
[  304s]         """Check that niimg is a proper 3D/4D niimg. Turn filenames into objects.
[  304s]     
[  304s]         Parameters
[  304s]         ----------
[  304s]         niimg : Niimg-like object
[  304s]             See http://nilearn.github.io/manipulating_images/input_output.html
[  304s]             If niimg is a string, consider it as a path to Nifti image and
[  304s]             call nibabel.load on it. The '~' symbol is expanded to the user home
[  304s]             folder.
[  304s]             If it is an object, check if the affine attribute present and that
[  304s]             nilearn.image.get_data returns a result, raise TypeError otherwise.
[  304s]     
[  304s]         ensure_ndim : integer {3, 4}, optional
[  304s]             Indicate the dimensionality of the expected niimg. An
[  304s]             error is raised if the niimg is of another dimensionality.
[  304s]     
[  304s]         atleast_4d : boolean, optional
[  304s]             Indicates if a 3d image should be turned into a single-scan 4d niimg.
[  304s]             Default=False.
[  304s]     
[  304s]         dtype : {None, dtype, "auto"}, optional
[  304s]             Data type toward which the data should be converted. If "auto", the
[  304s]             data will be converted to int32 if dtype is discrete and float32 if it
[  304s]             is continuous. If None, data will not be converted to a new data type.
[  304s]             Default=None.
[  304s]     
[  304s]         return_iterator : boolean, optional
[  304s]             Returns an iterator on the content of the niimg file input.
[  304s]             Default=False.
[  304s]     
[  304s]         wildcards : boolean, optional
[  304s]             Use niimg as a regular expression to get a list of matching input
[  304s]             filenames.
[  304s]             If multiple files match, the returned list is sorted using an ascending
[  304s]             order.
[  304s]             If no file matches the regular expression, a ValueError exception is
[  304s]             raised.
[  304s]             Default=True.
[  304s]     
[  304s]         Returns
[  304s]         -------
[  304s]         result : 3D/4D Niimg-like object
[  304s]             Result can be nibabel.Nifti1Image or the input, as-is. It is guaranteed
[  304s]             that the returned object has an affine attribute and that its data can
[  304s]             be retrieved with nilearn.image.get_data.
[  304s]     
[  304s]         Notes
[  304s]         -----
[  304s]         In nilearn, special care has been taken to make image manipulation easy.
[  304s]         This method is a kind of pre-requisite for any data processing method in
[  304s]         nilearn because it checks if data have a correct format and loads them if
[  304s]         necessary.
[  304s]     
[  304s]         Its application is idempotent.
[  304s]     
[  304s]         See also
[  304s]         --------
[  304s]             _iter_check_niimg, check_niimg_3d, check_niimg_4d
[  304s]     
[  304s]         """
[  304s]         from ..image import new_img_like  # avoid circular imports
[  304s]     
[  304s]         if isinstance(niimg, str):
[  304s]             if wildcards and ni.EXPAND_PATH_WILDCARDS:
[  304s]                 # Ascending sorting + expand user path
[  304s]                 filenames = sorted(glob.glob(os.path.expanduser(niimg)))
[  304s]     
[  304s]                 # processing filenames matching globbing expression
[  304s]                 if len(filenames) >= 1 and glob.has_magic(niimg):
[  304s]                     niimg = filenames  # iterable case
[  304s]                 # niimg is an existing filename
[  304s]                 elif [niimg] == filenames:
[  304s]                     niimg = filenames[0]
[  304s]                 # No files found by glob
[  304s]                 elif glob.has_magic(niimg):
[  304s]                     # No files matching the glob expression, warn the user
[  304s]                     message = ("No files matching the entered niimg expression: "
[  304s]                                 "'%s'.\n You may have left wildcards usage "
[  304s]                                "activated: please set the global constant "
[  304s]                                "'nilearn.EXPAND_PATH_WILDCARDS' to False to "
[  304s]                                "deactivate this behavior.") % niimg
[  304s]                     raise ValueError(message)
[  304s]                 else:
[  304s] >                   raise ValueError("File not found: '%s'" % niimg)
[  304s] E                   ValueError: File not found: '/tmp/nilearn_h4rbgbk1.nii'
[  304s] 
[  304s] nilearn/_utils/niimg_conversions.py:271: ValueError
```


### Steps and code to reproduce bug

Run tests on aarch64. 
Tested with openSUSE Tumbleweed aarch64 (and armv7, armv6).

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -179,10 +179,7 @@ def write_tmp_imgs(*imgs, **kwargs):
         finally:
             # Ensure all created files are removed
             for filename in filenames:
-                try:
-                    os.remove(filename)
-                except OSError:
-                    pass
+                os.remove(filename)
     else:  # No-op
         if len(imgs) == 1:
             yield imgs[0]

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -152,7 +152,7 @@ def write_tmp_imgs(*imgs, **kwargs):
     create_files = kwargs.get("create_files", True)
     use_wildcards = kwargs.get("use_wildcards", False)
 
-    prefix = "nilearn_{}_".format(os.getpid())
+    prefix = "nilearn_"
     suffix = ".nii"
 
     if create_files:
@@ -161,23 +161,25 @@ def write_tmp_imgs(*imgs, **kwargs):
             with warnings.catch_warnings():
                 warnings.simplefilter("ignore", RuntimeWarning)
                 for img in imgs:
-                    fd, path = tempfile.mkstemp(prefix=prefix,
-                                                suffix=suffix,
-                                                dir=None)
-                    filenames.append(path)
-                    img.to_filename(path)
+                    fd, filename = tempfile.mkstemp(prefix=prefix,
+                                                    suffix=suffix,
+                                                    dir=None)
+                    filenames.append((fd, filename))
+                    img.to_filename(filename)
                     del img
 
                 if use_wildcards:
                     yield prefix + "*" + suffix
                 else:
                     if len(imgs) == 1:
-                        yield filenames[0]
+                        yield filenames[0][1]
                     else:
-                        yield filenames
+                        yield [f[1] for f in filenames]
         finally:
             # Ensure all created files are removed
-            for filename in filenames:
+            for temp in filenames:
+                fd, filename = temp
+                os.close(fd)
                 os.remove(filename)
     else:  # No-op
         if len(imgs) == 1:

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -177,9 +177,20 @@ def write_tmp_imgs(*imgs, **kwargs):
                     else:
                         yield filenames
         finally:
+            failures = []
             # Ensure all created files are removed
             for filename in filenames:
-                os.remove(filename)
+                try:
+                    os.remove(filename)
+                except FileNotFoundError:
+                    # ok, file already removed
+                    pass
+                except OSError as e:
+                    # problem eg permission, or open file descriptor
+                    failures.append(e)
+            if failures:
+                raise ValueError(functools.reduce(lambda x,y: x+'\n'+y,
+                                    [str(e) for e in failures]))
     else:  # No-op
         if len(imgs) == 1:
             yield imgs[0]

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -189,8 +189,11 @@ def write_tmp_imgs(*imgs, **kwargs):
                     # problem eg permission, or open file descriptor
                     failures.append(e)
             if failures:
-                raise ValueError(functools.reduce(lambda x,y: x+'\n'+y,
-                                    [str(e) for e in failures]))
+                raise OSError(
+                    "The following files could not be removed:\n".format(
+                        "\n".join(str(e) for e in failures)
+                    )
+                )
     else:  # No-op
         if len(imgs) == 1:
             yield imgs[0]

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -164,7 +164,8 @@ def write_tmp_imgs(*imgs, **kwargs):
                     fd, filename = tempfile.mkstemp(prefix=prefix,
                                                     suffix=suffix,
                                                     dir=None)
-                    filenames.append((fd, filename))
+                    os.close(fd)
+                    filenames.append(filename)
                     img.to_filename(filename)
                     del img
 
@@ -172,15 +173,16 @@ def write_tmp_imgs(*imgs, **kwargs):
                     yield prefix + "*" + suffix
                 else:
                     if len(imgs) == 1:
-                        yield filenames[0][1]
+                        yield filenames[0]
                     else:
-                        yield [f[1] for f in filenames]
+                        yield filenames
         finally:
             # Ensure all created files are removed
-            for temp in filenames:
-                fd, filename = temp
-                os.close(fd)
-                os.remove(filename)
+            for filename in filenames:
+                try:
+                    os.remove(filename)
+                except OSError:
+                    pass
     else:  # No-op
         if len(imgs) == 1:
             yield imgs[0]

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -152,7 +152,7 @@ def write_tmp_imgs(*imgs, **kwargs):
     create_files = kwargs.get("create_files", True)
     use_wildcards = kwargs.get("use_wildcards", False)
 
-    prefix = "nilearn_"
+    prefix = "nilearn_{}_".format(os.getpid())
     suffix = ".nii"
 
     if create_files:

--- a/nilearn/_utils/testing.py
+++ b/nilearn/_utils/testing.py
@@ -161,11 +161,11 @@ def write_tmp_imgs(*imgs, **kwargs):
             with warnings.catch_warnings():
                 warnings.simplefilter("ignore", RuntimeWarning)
                 for img in imgs:
-                    filename = tempfile.mktemp(prefix=prefix,
-                                               suffix=suffix,
-                                               dir=None)
-                    filenames.append(filename)
-                    img.to_filename(filename)
+                    fd, path = tempfile.mkstemp(prefix=prefix,
+                                                suffix=suffix,
+                                                dir=None)
+                    filenames.append(path)
+                    img.to_filename(path)
                     del img
 
                 if use_wildcards:

--- a/nilearn/surface/tests/test_surface.py
+++ b/nilearn/surface/tests/test_surface.py
@@ -74,9 +74,11 @@ def test_load_surf_data_array():
     assert_array_equal(load_surf_data(data_squeeze), np.zeros((20, 3)))
 
 
-def test_load_surf_data_file_nii_gii():
+def test_load_surf_data_file_nii_gii(tmp_path):
     # test loading of fake data from gifti file
-    fd_gii, filename_gii = tempfile.mkstemp(suffix='.gii')
+    fd_gii, filename_gii = tempfile.mkstemp(suffix='.gii',
+                                            dir=tmp_path)
+    os.close(fd_gii)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray = gifti.GiftiDataArray(data=np.zeros((20, )))
     else:
@@ -87,30 +89,32 @@ def test_load_surf_data_file_nii_gii():
     gii = gifti.GiftiImage(darrays=[darray])
     gifti.write(gii, filename_gii)
     assert_array_equal(load_surf_data(filename_gii), np.zeros((20, )))
-    os.close(fd_gii)
     os.remove(filename_gii)
 
     # test loading of data from empty gifti file
-    fd_empty, filename_gii_empty = tempfile.mkstemp(suffix='.gii')
+    fd_empty, filename_gii_empty = tempfile.mkstemp(suffix='.gii',
+                                                    dir=tmp_path)
+    os.close(fd_empty)
     gii_empty = gifti.GiftiImage()
     gifti.write(gii_empty, filename_gii_empty)
     with pytest.raises(ValueError,
                        match='must contain at least one data array'
                        ):
         load_surf_data(filename_gii_empty)
-    os.close(fd_empty)
     os.remove(filename_gii_empty)
 
     # test loading of fake data from nifti file
-    fd_gii2, filename_nii = tempfile.mkstemp(suffix='.nii')
-    fd_niigz, filename_niigz = tempfile.mkstemp(suffix='.nii.gz')
+    fd_gii2, filename_nii = tempfile.mkstemp(suffix='.nii',
+                                             dir=tmp_path)
+    os.close(fd_gii2)
+    fd_niigz, filename_niigz = tempfile.mkstemp(suffix='.nii.gz',
+                                                dir=tmp_path)
+    os.close(fd_niigz)
     nii = nb.Nifti1Image(np.zeros((20, )), affine=None)
     nb.save(nii, filename_nii)
     nb.save(nii, filename_niigz)
     assert_array_equal(load_surf_data(filename_nii), np.zeros((20, )))
     assert_array_equal(load_surf_data(filename_niigz), np.zeros((20, )))
-    os.close(fd_gii2)
-    os.close(fd_niigz)
     os.remove(filename_nii)
     os.remove(filename_niigz)
 
@@ -132,7 +136,7 @@ def test_load_surf_data_gii_gz():
     assert isinstance(gii, gifti.GiftiImage)
 
 
-def test_load_surf_data_file_freesurfer():
+def test_load_surf_data_file_freesurfer(tmp_path):
     # test loading of fake data from sulc and thickness files
     # using load_surf_data.
     # We test load_surf_data by creating fake data with function
@@ -140,16 +144,18 @@ def test_load_surf_data_file_freesurfer():
     # version is recent with nibabel >= 2.1.0
     if LooseVersion(nb.__version__) >= LooseVersion('2.1.0'):
         data = np.zeros((20, ))
-        fd_sulc, filename_sulc = tempfile.mkstemp(suffix='.sulc')
+        fd_sulc, filename_sulc = tempfile.mkstemp(suffix='.sulc',
+                                                  dir=tmp_path)
+        os.close(fd_sulc)
         nb.freesurfer.io.write_morph_data(filename_sulc, data)
         assert_array_equal(load_surf_data(filename_sulc), np.zeros((20, )))
-        os.close(fd_sulc)
         os.remove(filename_sulc)
 
-        fd_thick, filename_thick = tempfile.mkstemp(suffix='.thickness')
+        fd_thick, filename_thick = tempfile.mkstemp(suffix='.thickness',
+                                                    dir=tmp_path)
+        os.close(fd_thick)
         nb.freesurfer.io.write_morph_data(filename_thick, data)
         assert_array_equal(load_surf_data(filename_thick), np.zeros((20, )))
-        os.close(fd_thick)
         os.remove(filename_thick)
 
     # test loading of data from real label and annot files
@@ -170,18 +176,19 @@ def test_load_surf_data_file_freesurfer():
     del annot, annot_start, annot_end
 
 
-def test_load_surf_data_file_error():
+def test_load_surf_data_file_error(tmp_path):
     # test if files with unexpected suffixes raise errors
     data = np.zeros((20, ))
     wrong_suff = ['.vtk', '.obj', '.mnc', '.txt']
     for suff in wrong_suff:
-        fd, filename_wrong = tempfile.mkstemp(suffix=suff)
+        fd, filename_wrong = tempfile.mkstemp(suffix=suff,
+                                              dir=tmp_path)
+        os.close(fd)
         np.savetxt(filename_wrong, data)
         with pytest.raises(ValueError,
                            match='input type is not recognized'
                            ):
             load_surf_data(filename_wrong)
-        os.close(fd)
         os.remove(filename_wrong)
 
 
@@ -284,7 +291,7 @@ def test_load_surf_mesh_file_gii_gz():
     assert isinstance(faces, np.ndarray)
 
 
-def test_load_surf_mesh_file_gii():
+def test_load_surf_mesh_file_gii(tmp_path):
     # Test the loader `load_surf_mesh`
 
     # If nibabel is of older version we skip tests as nibabel does not
@@ -297,8 +304,9 @@ def test_load_surf_mesh_file_gii():
     mesh = generate_surf()
 
     # test if correct gii is loaded into correct list
-    fd_mesh, filename_gii_mesh = tempfile.mkstemp(suffix='.gii')
-
+    fd_mesh, filename_gii_mesh = tempfile.mkstemp(suffix='.gii',
+                                                  dir=tmp_path)
+    os.close(fd_mesh)
     coord_array = gifti.GiftiDataArray(data=mesh[0],
                                        intent=nb.nifti1.intent_codes[
                                            'NIFTI_INTENT_POINTSET'])
@@ -310,61 +318,68 @@ def test_load_surf_mesh_file_gii():
     gifti.write(gii, filename_gii_mesh)
     assert_array_equal(load_surf_mesh(filename_gii_mesh)[0], mesh[0])
     assert_array_equal(load_surf_mesh(filename_gii_mesh)[1], mesh[1])
-    os.close(fd_mesh)
     os.remove(filename_gii_mesh)
 
     # test if incorrect gii raises error
-    fd_no, filename_gii_mesh_no_point = tempfile.mkstemp(suffix='.gii')
+    fd_no, filename_gii_mesh_no_point = tempfile.mkstemp(suffix='.gii',
+                                                         dir=tmp_path)
+    os.close(fd_no)
     gifti.write(gifti.GiftiImage(darrays=[face_array, face_array]),
                 filename_gii_mesh_no_point)
     with pytest.raises(ValueError, match='NIFTI_INTENT_POINTSET'):
         load_surf_mesh(filename_gii_mesh_no_point)
-    os.close(fd_no)
     os.remove(filename_gii_mesh_no_point)
 
-    fd_face, filename_gii_mesh_no_face = tempfile.mkstemp(suffix='.gii')
+    fd_face, filename_gii_mesh_no_face = tempfile.mkstemp(suffix='.gii',
+                                                          dir=tmp_path)
+    os.close(fd_face)
     gifti.write(gifti.GiftiImage(darrays=[coord_array, coord_array]),
                 filename_gii_mesh_no_face)
     with pytest.raises(ValueError, match='NIFTI_INTENT_TRIANGLE'):
         load_surf_mesh(filename_gii_mesh_no_face)
-    os.close(fd_face)
     os.remove(filename_gii_mesh_no_face)
 
 
-def test_load_surf_mesh_file_freesurfer():
+def test_load_surf_mesh_file_freesurfer(tmp_path):
     mesh = generate_surf()
     for suff in ['.pial', '.inflated', '.white', '.orig', 'sphere']:
-        fd, filename_fs_mesh = tempfile.mkstemp(suffix=suff)
+        fd, filename_fs_mesh = tempfile.mkstemp(suffix=suff,
+                                                dir=tmp_path)
+        os.close(fd)
         nb.freesurfer.write_geometry(filename_fs_mesh, mesh[0], mesh[1])
         assert len(load_surf_mesh(filename_fs_mesh)) == 2
         assert_array_almost_equal(load_surf_mesh(filename_fs_mesh)[0],
                                   mesh[0])
         assert_array_almost_equal(load_surf_mesh(filename_fs_mesh)[1],
                                   mesh[1])
-        os.close(fd)
         os.remove(filename_fs_mesh)
 
 
-def test_load_surf_mesh_file_error():
+def test_load_surf_mesh_file_error(tmp_path):
     # test if files with unexpected suffixes raise errors
     mesh = generate_surf()
     wrong_suff = ['.vtk', '.obj', '.mnc', '.txt']
     for suff in wrong_suff:
-        fd, filename_wrong = tempfile.mkstemp(suffix=suff)
+        fd, filename_wrong = tempfile.mkstemp(suffix=suff,
+                                              dir=tmp_path)
+        os.close(fd)
         nb.freesurfer.write_geometry(filename_wrong, mesh[0], mesh[1])
         with pytest.raises(ValueError,
                            match='input type is not recognized'
                            ):
             load_surf_mesh(filename_wrong)
-        os.close(fd)
         os.remove(filename_wrong)
 
 
-def test_load_surf_mesh_file_glob():
+def test_load_surf_mesh_file_glob(tmp_path):
     mesh = generate_surf()
-    fd1, fname1 = tempfile.mkstemp(suffix='.pial')
+    fd1, fname1 = tempfile.mkstemp(suffix='.pial',
+                                   dir=tmp_path)
+    os.close(fd1)
     nb.freesurfer.write_geometry(fname1, mesh[0], mesh[1])
-    fd2, fname2 = tempfile.mkstemp(suffix='.pial')
+    fd2, fname2 = tempfile.mkstemp(suffix='.pial',
+                                   dir=tmp_path)
+    os.close(fd2)
     nb.freesurfer.write_geometry(fname2, mesh[0], mesh[1])
 
     with pytest.raises(ValueError, match='More than one file matching path'):
@@ -376,22 +391,20 @@ def test_load_surf_mesh_file_glob():
     assert len(load_surf_mesh(fname1)) == 2
     assert_array_almost_equal(load_surf_mesh(fname1)[0], mesh[0])
     assert_array_almost_equal(load_surf_mesh(fname1)[1], mesh[1])
-
-    os.close(fd1)
-    os.close(fd2)
     os.remove(fname1)
     os.remove(fname2)
 
 
-def test_load_surf_data_file_glob():
+def test_load_surf_data_file_glob(tmp_path):
 
     data2D = np.ones((20, 3))
     fnames = []
     for f in range(3):
         fd, filename = tempfile.mkstemp(prefix='glob_%s_' % f,
-                                        suffix='.gii')
-        fnames.append(filename)
+                                        suffix='.gii',
+                                        dir=tmp_path)
         os.close(fd)
+        fnames.append(filename)
         data2D[:, f] *= f
         if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
             darray = gifti.GiftiDataArray(data=data2D[:, f])
@@ -410,9 +423,10 @@ def test_load_surf_data_file_glob():
 
     # make one more gii file that has more than one dimension
     fd, filename = tempfile.mkstemp(prefix='glob_3_',
-                                    suffix='.gii')
-    fnames.append(filename)
+                                    suffix='.gii',
+                                    dir=tmp_path)
     os.close(fd)
+    fnames.append(filename)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray1 = gifti.GiftiDataArray(data=np.ones((20, )))
         darray2 = gifti.GiftiDataArray(data=np.ones((20, )))
@@ -435,9 +449,10 @@ def test_load_surf_data_file_glob():
 
     # make one more gii file that has a different shape in axis=0
     fd, filename = tempfile.mkstemp(prefix='glob_4_',
-                                    suffix='.gii')
-    fnames.append(filename)
+                                    suffix='.gii',
+                                    dir=tmp_path)
     os.close(fd)
+    fnames.append(filename)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray = gifti.GiftiDataArray(data=np.ones((15, 1)))
     else:

--- a/nilearn/surface/tests/test_surface.py
+++ b/nilearn/surface/tests/test_surface.py
@@ -76,7 +76,7 @@ def test_load_surf_data_array():
 
 def test_load_surf_data_file_nii_gii():
     # test loading of fake data from gifti file
-    filename_gii = tempfile.mktemp(suffix='.gii')
+    fd_gii, filename_gii = tempfile.mkstemp(suffix='.gii')
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray = gifti.GiftiDataArray(data=np.zeros((20, )))
     else:
@@ -90,7 +90,7 @@ def test_load_surf_data_file_nii_gii():
     os.remove(filename_gii)
 
     # test loading of data from empty gifti file
-    filename_gii_empty = tempfile.mktemp(suffix='.gii')
+    fd_empty, filename_gii_empty = tempfile.mkstemp(suffix='.gii')
     gii_empty = gifti.GiftiImage()
     gifti.write(gii_empty, filename_gii_empty)
     with pytest.raises(ValueError,
@@ -100,8 +100,8 @@ def test_load_surf_data_file_nii_gii():
     os.remove(filename_gii_empty)
 
     # test loading of fake data from nifti file
-    filename_nii = tempfile.mktemp(suffix='.nii')
-    filename_niigz = tempfile.mktemp(suffix='.nii.gz')
+    fd_gii2, filename_nii = tempfile.mkstemp(suffix='.nii')
+    fd_niigz, filename_niigz = tempfile.mkstemp(suffix='.nii.gz')
     nii = nb.Nifti1Image(np.zeros((20, )), affine=None)
     nb.save(nii, filename_nii)
     nb.save(nii, filename_niigz)
@@ -136,12 +136,12 @@ def test_load_surf_data_file_freesurfer():
     # version is recent with nibabel >= 2.1.0
     if LooseVersion(nb.__version__) >= LooseVersion('2.1.0'):
         data = np.zeros((20, ))
-        filename_sulc = tempfile.mktemp(suffix='.sulc')
+        fd_sulc, filename_sulc = tempfile.mkstemp(suffix='.sulc')
         nb.freesurfer.io.write_morph_data(filename_sulc, data)
         assert_array_equal(load_surf_data(filename_sulc), np.zeros((20, )))
         os.remove(filename_sulc)
 
-        filename_thick = tempfile.mktemp(suffix='.thickness')
+        fd_thick, filename_thick = tempfile.mkstemp(suffix='.thickness')
         nb.freesurfer.io.write_morph_data(filename_thick, data)
         assert_array_equal(load_surf_data(filename_thick), np.zeros((20, )))
         os.remove(filename_thick)
@@ -169,7 +169,7 @@ def test_load_surf_data_file_error():
     data = np.zeros((20, ))
     wrong_suff = ['.vtk', '.obj', '.mnc', '.txt']
     for suff in wrong_suff:
-        filename_wrong = tempfile.mktemp(suffix=suff)
+        fd, filename_wrong = tempfile.mkstemp(suffix=suff)
         np.savetxt(filename_wrong, data)
         with pytest.raises(ValueError,
                            match='input type is not recognized'
@@ -290,7 +290,7 @@ def test_load_surf_mesh_file_gii():
     mesh = generate_surf()
 
     # test if correct gii is loaded into correct list
-    filename_gii_mesh = tempfile.mktemp(suffix='.gii')
+    fd_mesh, filename_gii_mesh = tempfile.mkstemp(suffix='.gii')
 
     coord_array = gifti.GiftiDataArray(data=mesh[0],
                                        intent=nb.nifti1.intent_codes[
@@ -306,14 +306,14 @@ def test_load_surf_mesh_file_gii():
     os.remove(filename_gii_mesh)
 
     # test if incorrect gii raises error
-    filename_gii_mesh_no_point = tempfile.mktemp(suffix='.gii')
+    fd_no, filename_gii_mesh_no_point = tempfile.mkstemp(suffix='.gii')
     gifti.write(gifti.GiftiImage(darrays=[face_array, face_array]),
                 filename_gii_mesh_no_point)
     with pytest.raises(ValueError, match='NIFTI_INTENT_POINTSET'):
         load_surf_mesh(filename_gii_mesh_no_point)
     os.remove(filename_gii_mesh_no_point)
 
-    filename_gii_mesh_no_face = tempfile.mktemp(suffix='.gii')
+    fd_face, filename_gii_mesh_no_face = tempfile.mkstemp(suffix='.gii')
     gifti.write(gifti.GiftiImage(darrays=[coord_array, coord_array]),
                 filename_gii_mesh_no_face)
     with pytest.raises(ValueError, match='NIFTI_INTENT_TRIANGLE'):
@@ -324,7 +324,7 @@ def test_load_surf_mesh_file_gii():
 def test_load_surf_mesh_file_freesurfer():
     mesh = generate_surf()
     for suff in ['.pial', '.inflated', '.white', '.orig', 'sphere']:
-        filename_fs_mesh = tempfile.mktemp(suffix=suff)
+        fd, filename_fs_mesh = tempfile.mkstemp(suffix=suff)
         nb.freesurfer.write_geometry(filename_fs_mesh, mesh[0], mesh[1])
         assert len(load_surf_mesh(filename_fs_mesh)) == 2
         assert_array_almost_equal(load_surf_mesh(filename_fs_mesh)[0],
@@ -339,7 +339,7 @@ def test_load_surf_mesh_file_error():
     mesh = generate_surf()
     wrong_suff = ['.vtk', '.obj', '.mnc', '.txt']
     for suff in wrong_suff:
-        filename_wrong = tempfile.mktemp(suffix=suff)
+        fd, filename_wrong = tempfile.mkstemp(suffix=suff)
         nb.freesurfer.write_geometry(filename_wrong, mesh[0], mesh[1])
         with pytest.raises(ValueError,
                            match='input type is not recognized'
@@ -350,9 +350,9 @@ def test_load_surf_mesh_file_error():
 
 def test_load_surf_mesh_file_glob():
     mesh = generate_surf()
-    fname1 = tempfile.mktemp(suffix='.pial')
+    fd1, fname1 = tempfile.mkstemp(suffix='.pial')
     nb.freesurfer.write_geometry(fname1, mesh[0], mesh[1])
-    fname2 = tempfile.mktemp(suffix='.pial')
+    fd2, fname2 = tempfile.mkstemp(suffix='.pial')
     nb.freesurfer.write_geometry(fname2, mesh[0], mesh[1])
 
     with pytest.raises(ValueError, match='More than one file matching path'):
@@ -374,7 +374,9 @@ def test_load_surf_data_file_glob():
     data2D = np.ones((20, 3))
     fnames = []
     for f in range(3):
-        fnames.append(tempfile.mktemp(prefix='glob_%s_' % f, suffix='.gii'))
+        fd, filename = tempfile.mkstemp(prefix='glob_%s_' % f,
+                                        suffix='.gii')
+        fnames.append(filename)
         data2D[:, f] *= f
         if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
             darray = gifti.GiftiDataArray(data=data2D[:, f])
@@ -392,7 +394,9 @@ def test_load_surf_data_file_glob():
     )
 
     # make one more gii file that has more than one dimension
-    fnames.append(tempfile.mktemp(prefix='glob_3_', suffix='.gii'))
+    fd, filename = tempfile.mkstemp(prefix='glob_3_',
+                                    suffix='.gii')
+    fnames.append(filename)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray1 = gifti.GiftiDataArray(data=np.ones((20, )))
         darray2 = gifti.GiftiDataArray(data=np.ones((20, )))
@@ -414,7 +418,9 @@ def test_load_surf_data_file_glob():
                                                    "glob*.gii")), data2D)
 
     # make one more gii file that has a different shape in axis=0
-    fnames.append(tempfile.mktemp(prefix='glob_4_', suffix='.gii'))
+    fd, filename = tempfile.mkstemp(prefix='glob_4_',
+                                    suffix='.gii')
+    fnames.append(filename)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray = gifti.GiftiDataArray(data=np.ones((15, 1)))
     else:

--- a/nilearn/surface/tests/test_surface.py
+++ b/nilearn/surface/tests/test_surface.py
@@ -77,7 +77,7 @@ def test_load_surf_data_array():
 def test_load_surf_data_file_nii_gii(tmp_path):
     # test loading of fake data from gifti file
     fd_gii, filename_gii = tempfile.mkstemp(suffix='.gii',
-                                            dir=tmp_path)
+                                            dir=str(tmp_path))
     os.close(fd_gii)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray = gifti.GiftiDataArray(data=np.zeros((20, )))
@@ -93,7 +93,7 @@ def test_load_surf_data_file_nii_gii(tmp_path):
 
     # test loading of data from empty gifti file
     fd_empty, filename_gii_empty = tempfile.mkstemp(suffix='.gii',
-                                                    dir=tmp_path)
+                                                    dir=str(tmp_path))
     os.close(fd_empty)
     gii_empty = gifti.GiftiImage()
     gifti.write(gii_empty, filename_gii_empty)
@@ -105,10 +105,10 @@ def test_load_surf_data_file_nii_gii(tmp_path):
 
     # test loading of fake data from nifti file
     fd_gii2, filename_nii = tempfile.mkstemp(suffix='.nii',
-                                             dir=tmp_path)
+                                             dir=str(tmp_path))
     os.close(fd_gii2)
     fd_niigz, filename_niigz = tempfile.mkstemp(suffix='.nii.gz',
-                                                dir=tmp_path)
+                                                dir=str(tmp_path))
     os.close(fd_niigz)
     nii = nb.Nifti1Image(np.zeros((20, )), affine=None)
     nb.save(nii, filename_nii)
@@ -145,14 +145,14 @@ def test_load_surf_data_file_freesurfer(tmp_path):
     if LooseVersion(nb.__version__) >= LooseVersion('2.1.0'):
         data = np.zeros((20, ))
         fd_sulc, filename_sulc = tempfile.mkstemp(suffix='.sulc',
-                                                  dir=tmp_path)
+                                                  dir=str(tmp_path))
         os.close(fd_sulc)
         nb.freesurfer.io.write_morph_data(filename_sulc, data)
         assert_array_equal(load_surf_data(filename_sulc), np.zeros((20, )))
         os.remove(filename_sulc)
 
         fd_thick, filename_thick = tempfile.mkstemp(suffix='.thickness',
-                                                    dir=tmp_path)
+                                                    dir=str(tmp_path))
         os.close(fd_thick)
         nb.freesurfer.io.write_morph_data(filename_thick, data)
         assert_array_equal(load_surf_data(filename_thick), np.zeros((20, )))
@@ -182,7 +182,7 @@ def test_load_surf_data_file_error(tmp_path):
     wrong_suff = ['.vtk', '.obj', '.mnc', '.txt']
     for suff in wrong_suff:
         fd, filename_wrong = tempfile.mkstemp(suffix=suff,
-                                              dir=tmp_path)
+                                              dir=str(tmp_path))
         os.close(fd)
         np.savetxt(filename_wrong, data)
         with pytest.raises(ValueError,
@@ -305,7 +305,7 @@ def test_load_surf_mesh_file_gii(tmp_path):
 
     # test if correct gii is loaded into correct list
     fd_mesh, filename_gii_mesh = tempfile.mkstemp(suffix='.gii',
-                                                  dir=tmp_path)
+                                                  dir=str(tmp_path))
     os.close(fd_mesh)
     coord_array = gifti.GiftiDataArray(data=mesh[0],
                                        intent=nb.nifti1.intent_codes[
@@ -322,7 +322,7 @@ def test_load_surf_mesh_file_gii(tmp_path):
 
     # test if incorrect gii raises error
     fd_no, filename_gii_mesh_no_point = tempfile.mkstemp(suffix='.gii',
-                                                         dir=tmp_path)
+                                                         dir=str(tmp_path))
     os.close(fd_no)
     gifti.write(gifti.GiftiImage(darrays=[face_array, face_array]),
                 filename_gii_mesh_no_point)
@@ -331,7 +331,7 @@ def test_load_surf_mesh_file_gii(tmp_path):
     os.remove(filename_gii_mesh_no_point)
 
     fd_face, filename_gii_mesh_no_face = tempfile.mkstemp(suffix='.gii',
-                                                          dir=tmp_path)
+                                                          dir=str(tmp_path))
     os.close(fd_face)
     gifti.write(gifti.GiftiImage(darrays=[coord_array, coord_array]),
                 filename_gii_mesh_no_face)
@@ -344,7 +344,7 @@ def test_load_surf_mesh_file_freesurfer(tmp_path):
     mesh = generate_surf()
     for suff in ['.pial', '.inflated', '.white', '.orig', 'sphere']:
         fd, filename_fs_mesh = tempfile.mkstemp(suffix=suff,
-                                                dir=tmp_path)
+                                                dir=str(tmp_path))
         os.close(fd)
         nb.freesurfer.write_geometry(filename_fs_mesh, mesh[0], mesh[1])
         assert len(load_surf_mesh(filename_fs_mesh)) == 2
@@ -361,7 +361,7 @@ def test_load_surf_mesh_file_error(tmp_path):
     wrong_suff = ['.vtk', '.obj', '.mnc', '.txt']
     for suff in wrong_suff:
         fd, filename_wrong = tempfile.mkstemp(suffix=suff,
-                                              dir=tmp_path)
+                                              dir=str(tmp_path))
         os.close(fd)
         nb.freesurfer.write_geometry(filename_wrong, mesh[0], mesh[1])
         with pytest.raises(ValueError,
@@ -374,11 +374,11 @@ def test_load_surf_mesh_file_error(tmp_path):
 def test_load_surf_mesh_file_glob(tmp_path):
     mesh = generate_surf()
     fd1, fname1 = tempfile.mkstemp(suffix='.pial',
-                                   dir=tmp_path)
+                                   dir=str(tmp_path))
     os.close(fd1)
     nb.freesurfer.write_geometry(fname1, mesh[0], mesh[1])
     fd2, fname2 = tempfile.mkstemp(suffix='.pial',
-                                   dir=tmp_path)
+                                   dir=str(tmp_path))
     os.close(fd2)
     nb.freesurfer.write_geometry(fname2, mesh[0], mesh[1])
 
@@ -402,7 +402,7 @@ def test_load_surf_data_file_glob(tmp_path):
     for f in range(3):
         fd, filename = tempfile.mkstemp(prefix='glob_%s_' % f,
                                         suffix='.gii',
-                                        dir=tmp_path)
+                                        dir=str(tmp_path))
         os.close(fd)
         fnames.append(filename)
         data2D[:, f] *= f
@@ -424,7 +424,7 @@ def test_load_surf_data_file_glob(tmp_path):
     # make one more gii file that has more than one dimension
     fd, filename = tempfile.mkstemp(prefix='glob_3_',
                                     suffix='.gii',
-                                    dir=tmp_path)
+                                    dir=str(tmp_path))
     os.close(fd)
     fnames.append(filename)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
@@ -450,7 +450,7 @@ def test_load_surf_data_file_glob(tmp_path):
     # make one more gii file that has a different shape in axis=0
     fd, filename = tempfile.mkstemp(prefix='glob_4_',
                                     suffix='.gii',
-                                    dir=tmp_path)
+                                    dir=str(tmp_path))
     os.close(fd)
     fnames.append(filename)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):

--- a/nilearn/surface/tests/test_surface.py
+++ b/nilearn/surface/tests/test_surface.py
@@ -87,6 +87,7 @@ def test_load_surf_data_file_nii_gii():
     gii = gifti.GiftiImage(darrays=[darray])
     gifti.write(gii, filename_gii)
     assert_array_equal(load_surf_data(filename_gii), np.zeros((20, )))
+    os.close(fd_gii)
     os.remove(filename_gii)
 
     # test loading of data from empty gifti file
@@ -97,6 +98,7 @@ def test_load_surf_data_file_nii_gii():
                        match='must contain at least one data array'
                        ):
         load_surf_data(filename_gii_empty)
+    os.close(fd_empty)
     os.remove(filename_gii_empty)
 
     # test loading of fake data from nifti file
@@ -107,6 +109,8 @@ def test_load_surf_data_file_nii_gii():
     nb.save(nii, filename_niigz)
     assert_array_equal(load_surf_data(filename_nii), np.zeros((20, )))
     assert_array_equal(load_surf_data(filename_niigz), np.zeros((20, )))
+    os.close(fd_gii2)
+    os.close(fd_niigz)
     os.remove(filename_nii)
     os.remove(filename_niigz)
 
@@ -139,11 +143,13 @@ def test_load_surf_data_file_freesurfer():
         fd_sulc, filename_sulc = tempfile.mkstemp(suffix='.sulc')
         nb.freesurfer.io.write_morph_data(filename_sulc, data)
         assert_array_equal(load_surf_data(filename_sulc), np.zeros((20, )))
+        os.close(fd_sulc)
         os.remove(filename_sulc)
 
         fd_thick, filename_thick = tempfile.mkstemp(suffix='.thickness')
         nb.freesurfer.io.write_morph_data(filename_thick, data)
         assert_array_equal(load_surf_data(filename_thick), np.zeros((20, )))
+        os.close(fd_thick)
         os.remove(filename_thick)
 
     # test loading of data from real label and annot files
@@ -175,6 +181,7 @@ def test_load_surf_data_file_error():
                            match='input type is not recognized'
                            ):
             load_surf_data(filename_wrong)
+        os.close(fd)
         os.remove(filename_wrong)
 
 
@@ -303,6 +310,7 @@ def test_load_surf_mesh_file_gii():
     gifti.write(gii, filename_gii_mesh)
     assert_array_equal(load_surf_mesh(filename_gii_mesh)[0], mesh[0])
     assert_array_equal(load_surf_mesh(filename_gii_mesh)[1], mesh[1])
+    os.close(fd_mesh)
     os.remove(filename_gii_mesh)
 
     # test if incorrect gii raises error
@@ -311,6 +319,7 @@ def test_load_surf_mesh_file_gii():
                 filename_gii_mesh_no_point)
     with pytest.raises(ValueError, match='NIFTI_INTENT_POINTSET'):
         load_surf_mesh(filename_gii_mesh_no_point)
+    os.close(fd_no)
     os.remove(filename_gii_mesh_no_point)
 
     fd_face, filename_gii_mesh_no_face = tempfile.mkstemp(suffix='.gii')
@@ -318,6 +327,7 @@ def test_load_surf_mesh_file_gii():
                 filename_gii_mesh_no_face)
     with pytest.raises(ValueError, match='NIFTI_INTENT_TRIANGLE'):
         load_surf_mesh(filename_gii_mesh_no_face)
+    os.close(fd_face)
     os.remove(filename_gii_mesh_no_face)
 
 
@@ -331,6 +341,7 @@ def test_load_surf_mesh_file_freesurfer():
                                   mesh[0])
         assert_array_almost_equal(load_surf_mesh(filename_fs_mesh)[1],
                                   mesh[1])
+        os.close(fd)
         os.remove(filename_fs_mesh)
 
 
@@ -345,6 +356,7 @@ def test_load_surf_mesh_file_error():
                            match='input type is not recognized'
                            ):
             load_surf_mesh(filename_wrong)
+        os.close(fd)
         os.remove(filename_wrong)
 
 
@@ -365,6 +377,8 @@ def test_load_surf_mesh_file_glob():
     assert_array_almost_equal(load_surf_mesh(fname1)[0], mesh[0])
     assert_array_almost_equal(load_surf_mesh(fname1)[1], mesh[1])
 
+    os.close(fd1)
+    os.close(fd2)
     os.remove(fname1)
     os.remove(fname2)
 
@@ -377,6 +391,7 @@ def test_load_surf_data_file_glob():
         fd, filename = tempfile.mkstemp(prefix='glob_%s_' % f,
                                         suffix='.gii')
         fnames.append(filename)
+        os.close(fd)
         data2D[:, f] *= f
         if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
             darray = gifti.GiftiDataArray(data=data2D[:, f])
@@ -397,6 +412,7 @@ def test_load_surf_data_file_glob():
     fd, filename = tempfile.mkstemp(prefix='glob_3_',
                                     suffix='.gii')
     fnames.append(filename)
+    os.close(fd)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray1 = gifti.GiftiDataArray(data=np.ones((20, )))
         darray2 = gifti.GiftiDataArray(data=np.ones((20, )))
@@ -421,6 +437,7 @@ def test_load_surf_data_file_glob():
     fd, filename = tempfile.mkstemp(prefix='glob_4_',
                                     suffix='.gii')
     fnames.append(filename)
+    os.close(fd)
     if LooseVersion(nb.__version__) > LooseVersion('2.0.2'):
         darray = gifti.GiftiDataArray(data=np.ones((15, 1)))
     else:

--- a/nilearn/tests/test_niimg_conversions.py
+++ b/nilearn/tests/test_niimg_conversions.py
@@ -357,7 +357,7 @@ def test_iter_check_niimgs(tmp_path):
     # Create a test file
     fd, filename = tempfile.mkstemp(prefix="nilearn_test",
                                     suffix=".nii",
-                                    dir=tmp_path)
+                                    dir=str(tmp_path))
     os.close(fd)
     img_4d.to_filename(filename)
     niimgs = list(_iter_check_niimg([filename]))
@@ -507,7 +507,7 @@ def test_repr_niimgs(tmp_path):
 
     # Add filename long enough to qualify for shortening
     fd, tmpimg1 = tempfile.mkstemp(suffix='_very_long.nii',
-                                   dir=tmp_path)
+                                   dir=str(tmp_path))
     os.close(fd)
     nibabel.save(img1, tmpimg1)
     assert (

--- a/nilearn/tests/test_niimg_conversions.py
+++ b/nilearn/tests/test_niimg_conversions.py
@@ -359,6 +359,7 @@ def test_iter_check_niimgs():
                                     suffix=".nii",
                                     dir=None)
     img_4d.to_filename(filename)
+    os.close(fd)
     niimgs = list(_iter_check_niimg([filename]))
     assert_array_equal(get_data(niimgs[0]),
                        get_data(_utils.check_niimg(img_4d)))
@@ -505,8 +506,9 @@ def test_repr_niimgs():
                 repr(shape), repr(affine))))
 
     # Add filename long enough to qualify for shortening
-    _, tmpimg1 = tempfile.mkstemp(suffix='_very_long.nii')
+    fd, tmpimg1 = tempfile.mkstemp(suffix='_very_long.nii')
     nibabel.save(img1, tmpimg1)
+    os.close(fd)
     assert (
         _utils._repr_niimgs(img1, shorten=False) ==
         ("%s('%s')" % (img1.__class__.__name__, img1.get_filename())))

--- a/nilearn/tests/test_niimg_conversions.py
+++ b/nilearn/tests/test_niimg_conversions.py
@@ -340,7 +340,7 @@ def test_check_niimg_wildcards():
     ni.EXPAND_PATH_WILDCARDS = True
 
 
-def test_iter_check_niimgs():
+def test_iter_check_niimgs(tmp_path):
     no_file_matching = "No files matching path: %s"
     affine = np.eye(4)
     img_4d = Nifti1Image(np.ones((10, 10, 10, 4)), affine)
@@ -357,9 +357,9 @@ def test_iter_check_niimgs():
     # Create a test file
     fd, filename = tempfile.mkstemp(prefix="nilearn_test",
                                     suffix=".nii",
-                                    dir=None)
-    img_4d.to_filename(filename)
+                                    dir=tmp_path)
     os.close(fd)
+    img_4d.to_filename(filename)
     niimgs = list(_iter_check_niimg([filename]))
     assert_array_equal(get_data(niimgs[0]),
                        get_data(_utils.check_niimg(img_4d)))
@@ -390,7 +390,7 @@ def test_iter_check_niimgs_memory():
                              for i in range(10)])
 
 
-def test_repr_niimgs():
+def test_repr_niimgs(tmp_path):
     # Tests with file path
     assert _utils._repr_niimgs("test") == "test"
     assert _utils._repr_niimgs("test", shorten=False) == "test"
@@ -506,9 +506,10 @@ def test_repr_niimgs():
                 repr(shape), repr(affine))))
 
     # Add filename long enough to qualify for shortening
-    fd, tmpimg1 = tempfile.mkstemp(suffix='_very_long.nii')
-    nibabel.save(img1, tmpimg1)
+    fd, tmpimg1 = tempfile.mkstemp(suffix='_very_long.nii',
+                                   dir=tmp_path)
     os.close(fd)
+    nibabel.save(img1, tmpimg1)
     assert (
         _utils._repr_niimgs(img1, shorten=False) ==
         ("%s('%s')" % (img1.__class__.__name__, img1.get_filename())))

--- a/nilearn/tests/test_niimg_conversions.py
+++ b/nilearn/tests/test_niimg_conversions.py
@@ -355,9 +355,9 @@ def test_iter_check_niimgs():
         list(_iter_check_niimg(nofile_path))
 
     # Create a test file
-    filename = tempfile.mktemp(prefix="nilearn_test",
-                               suffix=".nii",
-                               dir=None)
+    fd, filename = tempfile.mkstemp(prefix="nilearn_test",
+                                    suffix=".nii",
+                                    dir=None)
     img_4d.to_filename(filename)
     niimgs = list(_iter_check_niimg([filename]))
     assert_array_equal(get_data(niimgs[0]),

--- a/nilearn/tests/test_numpy_conversions.py
+++ b/nilearn/tests/test_numpy_conversions.py
@@ -236,7 +236,7 @@ def test_as_ndarray():
 def test_csv_to_array(tmp_path):
     # Create a phony CSV file
     fd, filename = tempfile.mkstemp(suffix='.csv',
-                                    dir=tmp_path)
+                                    dir=str(tmp_path))
     os.close(fd)
     try:
         with open(filename, mode='wt') as fp:

--- a/nilearn/tests/test_numpy_conversions.py
+++ b/nilearn/tests/test_numpy_conversions.py
@@ -233,7 +233,7 @@ def test_as_ndarray():
     pytest.raises(ValueError, as_ndarray, [], order="invalid")
 
 
-def test_csv_to_array(temp_path):
+def test_csv_to_array(tmp_path):
     # Create a phony CSV file
     fd, filename = tempfile.mkstemp(suffix='.csv',
                                     dir=tmp_path)

--- a/nilearn/tests/test_numpy_conversions.py
+++ b/nilearn/tests/test_numpy_conversions.py
@@ -235,7 +235,7 @@ def test_as_ndarray():
 
 def test_csv_to_array():
     # Create a phony CSV file
-    filename = tempfile.mktemp(suffix='.csv')
+    fd, filename = tempfile.mkstemp(suffix='.csv')
     try:
         with open(filename, mode='wt') as fp:
             fp.write('1.,2.,3.,4.,5.\n')

--- a/nilearn/tests/test_numpy_conversions.py
+++ b/nilearn/tests/test_numpy_conversions.py
@@ -233,9 +233,11 @@ def test_as_ndarray():
     pytest.raises(ValueError, as_ndarray, [], order="invalid")
 
 
-def test_csv_to_array():
+def test_csv_to_array(temp_path):
     # Create a phony CSV file
-    fd, filename = tempfile.mkstemp(suffix='.csv')
+    fd, filename = tempfile.mkstemp(suffix='.csv',
+                                    dir=tmp_path)
+    os.close(fd)
     try:
         with open(filename, mode='wt') as fp:
             fp.write('1.,2.,3.,4.,5.\n')
@@ -243,5 +245,4 @@ def test_csv_to_array():
                     np.asarray([1., 2., 3., 4., 5.]))
         pytest.raises(TypeError, csv_to_array, filename, delimiters='?!')
     finally:
-        os.close(fd)
         os.remove(filename)

--- a/nilearn/tests/test_numpy_conversions.py
+++ b/nilearn/tests/test_numpy_conversions.py
@@ -243,4 +243,5 @@ def test_csv_to_array():
                     np.asarray([1., 2., 3., 4., 5.]))
         pytest.raises(TypeError, csv_to_array, filename, delimiters='?!')
     finally:
+        os.close(fd)
         os.remove(filename)

